### PR TITLE
out_splunk: add new property "channel" to set channel identifier

### LINKED ISSUE #3314: Splunk output plugin: Data channel is missing
**Is your feature request related to a problem? Please describe.**
I already logged this story over in the fluent-bit-docs repo, but haven't heard a response in over a week, so I'm raising it here.
https://github.com/fluent/fluent-bit-docs/issues/496

The root of this problem is based on "indexer acknowledgement" in splunk. In a perfect world, the fluent-bit plugin would support indexer acknowledgement (I haven't looked at the source, so I don't know if it does), but this is helpful because it allows splunk to indicate (to fluent-bit) that the specified index isn't correct at the time of request (so fluent bit can log an error if a bad index is used). 

When indexer acknowledgement is enabled, the `x-splunk-request-channel` header is required to be specified as an HTTP header on the request to the HEC endpoint, but there's no way to specify that as part of the splunk plugin configuration. 

**Describe the solution you'd like**
I think it'd be possible to specify a static request channel string as part of the output section for the splunk plugin. 

**Describe alternatives you've considered**
The alternative is to just disable indexer acknowledgement. The trouble here is that fluent bit will believe everything is working correctly, but logs will be mysteriously missing. 

**Additional context**
n/a

### PR DESCRIPTION
**Pull request by nokute78**
disclaimer: I don't have splunk account. I tested dummy http server, so the output log may be invalid.
Waiting for testing by user.

Fixes #3314  

Refer to the splunk doc, we need to set `X-Splunk-Request-Channel` header.
This PR is to add new property `channel` to set the header.
https://docs.splunk.com/Documentation/SplunkCloud/latest/Data/FormateventsforHTTPEventCollector#Channel_identifier_header

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [X] Example configuration file for the change
- [X] Debug log output from testing the change

- [X] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

**Documentation**

- [ ] Documentation required for this feature

## Example Configuration

```
[INPUT]
    Name dummy
    Samples 1

[OUTPUT]
    Name splunk
    Http_User nanashi
    Channel bd3b4270-3c26-4127-a677-8643d44f3984
```

## Debug log output

disclaimer: I don't have splunk account. I tested dummy http server, so the output log may be invalid.

This is `nc -l 8088` outputs.
It reports `X-Splunk-Request-Channel: bd3b4270-3c26-4127-a677-8643d44f3984`.
```
$ nc -l 8088
POST /services/collector/event HTTP/1.1
Host: 127.0.0.1:8088
Content-Length: 54
User-Agent: Fluent-Bit
Authorization: Basic bmFuYXNoaTo=
X-Splunk-Request-Channel: bd3b4270-3c26-4127-a677-8643d44f3984

{"time":1617326861.963734,"event":{"message":"dummy"}}
```

Fluent-bit's log
```
$ ../bin/fluent-bit -c a.conf 
Fluent Bit v1.8.0
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2021/04/02 10:48:15] [ info] [engine] started (pid=8495)
[2021/04/02 10:48:15] [ info] [storage] version=1.1.1, initializing...
[2021/04/02 10:48:15] [ info] [storage] in-memory
[2021/04/02 10:48:15] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/04/02 10:48:15] [ info] [sp] stream processor started
^C[2021/04/02 10:48:21] [engine] caught signal (SIGINT)
[2021/04/02 10:48:21] [ warn] [engine] service will stop in 5 seconds
[2021/04/02 10:48:25] [ info] [engine] service stopped
[2021/04/02 10:48:25] [ warn] [engine] shutdown delayed, grace period has finished but some tasks are still running.
[2021/04/02 10:48:25] [ info] [task] dummy/dummy.0 has 1 pending task(s):
[2021/04/02 10:48:25] [ info] [task]   task_id=0 still running on route(s): splunk/splunk.0 
[2021/04/02 10:48:25] [ warn] [engine] service will stop in 5 seconds
[2021/04/02 10:48:30] [ info] [engine] service stopped
[2021/04/02 10:48:30] [ warn] [engine] shutdown delayed, grace period has finished but some tasks are still running.
[2021/04/02 10:48:30] [ info] [task] dummy/dummy.0 has 1 pending task(s):
[2021/04/02 10:48:30] [ info] [task]   task_id=0 still running on route(s): splunk/splunk.0 
[2021/04/02 10:48:30] [ warn] [engine] service will stop in 5 seconds
^C[2021/04/02 10:48:31] [engine] caught signal (SIGINT)
```

## Valgrind output

Note: valgrind reports leak, but it is not related this PR.
It reports `http_user` and `http_passwd` are leaked. (they are overwritten by flb_strdup)
```
$ valgrind --leak-check=full ../bin/fluent-bit -c a.conf 
==8319== Memcheck, a memory error detector
==8319== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==8319== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==8319== Command: ../bin/fluent-bit -c a.conf
==8319== 
Fluent Bit v1.8.0
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2021/04/02 10:40:09] [ info] [engine] started (pid=8319)
[2021/04/02 10:40:09] [ info] [storage] version=1.1.1, initializing...
[2021/04/02 10:40:09] [ info] [storage] in-memory
[2021/04/02 10:40:09] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/04/02 10:40:09] [ info] [sp] stream processor started
^C[2021/04/02 10:40:10] [engine] caught signal (SIGINT)
==8319== Warning: client switching stacks?  SP change: 0x57e4818 --> 0x4c57da0
==8319==          to suppress, use: --max-stackframe=12110456 or greater
==8319== Warning: client switching stacks?  SP change: 0x4c57d18 --> 0x57e4818
==8319==          to suppress, use: --max-stackframe=12110592 or greater
==8319== Warning: client switching stacks?  SP change: 0x57e4818 --> 0x4c57d18
==8319==          to suppress, use: --max-stackframe=12110592 or greater
==8319==          further instances of this message will not be shown.
[2021/04/02 10:40:10] [ warn] [engine] service will stop in 5 seconds
[2021/04/02 10:40:10] [error] [src/flb_http_client.c:1156 errno=32] Broken pipe
[2021/04/02 10:40:10] [ warn] [output:splunk:splunk.0] http_do=-1
[2021/04/02 10:40:10] [ warn] [engine] failed to flush chunk '8319-1617327609.959640794.flb', retry in 11 seconds: task_id=0, input=dummy.0 > output=splunk.0 (out_id=0)
[2021/04/02 10:40:14] [ info] [engine] service stopped
==8319== Warning: invalid file descriptor -1 in syscall close()
==8319== 
==8319== HEAP SUMMARY:
==8319==     in use at exit: 9 bytes in 2 blocks
==8319==   total heap usage: 353 allocs, 351 frees, 580,886 bytes allocated
==8319== 
==8319== 1 bytes in 1 blocks are definitely lost in loss record 1 of 2
==8319==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==8319==    by 0x209576: flb_malloc (flb_mem.h:62)
==8319==    by 0x20963F: flb_strdup (flb_str.h:37)
==8319==    by 0x209968: flb_splunk_conf_create (splunk_conf.c:98)
==8319==    by 0x208A0E: cb_splunk_init (splunk.c:37)
==8319==    by 0x1759B8: flb_output_init_all (flb_output.c:930)
==8319==    by 0x182276: flb_engine_start (flb_engine.c:516)
==8319==    by 0x16748C: flb_lib_worker (flb_lib.c:493)
==8319==    by 0x4864608: start_thread (pthread_create.c:477)
==8319==    by 0x4B10292: clone (clone.S:95)
==8319== 
==8319== 8 bytes in 1 blocks are definitely lost in loss record 2 of 2
==8319==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==8319==    by 0x209576: flb_malloc (flb_mem.h:62)
==8319==    by 0x20963F: flb_strdup (flb_str.h:37)
==8319==    by 0x209920: flb_splunk_conf_create (splunk_conf.c:92)
==8319==    by 0x208A0E: cb_splunk_init (splunk.c:37)
==8319==    by 0x1759B8: flb_output_init_all (flb_output.c:930)
==8319==    by 0x182276: flb_engine_start (flb_engine.c:516)
==8319==    by 0x16748C: flb_lib_worker (flb_lib.c:493)
==8319==    by 0x4864608: start_thread (pthread_create.c:477)
==8319==    by 0x4B10292: clone (clone.S:95)
==8319== 
==8319== LEAK SUMMARY:
==8319==    definitely lost: 9 bytes in 2 blocks
==8319==    indirectly lost: 0 bytes in 0 blocks
==8319==      possibly lost: 0 bytes in 0 blocks
==8319==    still reachable: 0 bytes in 0 blocks
==8319==         suppressed: 0 bytes in 0 blocks
==8319== 
==8319== For lists of detected and suppressed errors, rerun with: -s
==8319== ERROR SUMMARY: 2 errors from 2 contexts (suppressed: 0 from 0)
```

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

**Comment by nokute78** [comment]
Note: 
I created another PR #3319 to fix leak.
```
Note: valgrind reports leak, but it is not related this PR.
It reports http_user and http_passwd are leaked. (they are overwritten by flb_strdup)
```

**Comment by edsiper** [comment]
FYI: you can deploy Splunk in a docker container using the free tier: https://docs.splunk.com/Documentation/Splunk/8.1.3/Installation/DeployandrunSplunkEnterpriseinsideDockercontainers

**Comment by nokute78** [comment]
Thank you. I'll check it later.

**Comment by nokute78** [comment]
I tested splunk container image and it works!
https://splunk.github.io/docker-splunk/EXAMPLES.html#create-standalone-with-hec

## setting splunk
1. login splunk (access localhost:8000 by browser) 
2. **Setting** -> **Data inputs**
3. **HTTP Event Collector**  -> Actions **Edit**
4. Check **Enable indexer acknowledgement**

## Output
If `Channel` is not set, splunk returns error `400`.
```
$ ../bin/fluent-bit -c a.conf 
Fluent Bit v1.8.0
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2021/04/05 09:18:17] [ info] [engine] started (pid=28363)
[2021/04/05 09:18:17] [ info] [storage] version=1.1.1, initializing...
[2021/04/05 09:18:17] [ info] [storage] in-memory
[2021/04/05 09:18:17] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/04/05 09:18:17] [ info] [sp] stream processor started
[2021/04/05 09:18:22] [ warn] [output:splunk:splunk.0] http_status=400:
{"text":"Data channel is missing","code":10}
```

If `Channel` is set, splunk doesn't return error `400`.
```
$ ../bin/fluent-bit -c a.conf 
Fluent Bit v1.8.0
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2021/04/05 09:18:28] [ info] [engine] started (pid=28371)
[2021/04/05 09:18:28] [ info] [storage] version=1.1.1, initializing...
[2021/04/05 09:18:28] [ info] [storage] in-memory
[2021/04/05 09:18:28] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/04/05 09:18:28] [ info] [sp] stream processor started
^C[2021/04/05 09:18:37] [engine] caught signal (SIGINT)
[2021/04/05 09:18:37] [ warn] [engine] service will stop in 5 seconds
^C[2021/04/05 09:18:38] [engine] caught signal (SIGINT)
```

**Comment by nokute78** [comment]
CI passed!

File: plugins/out_splunk/splunk.c [hunk 1]
```diff
@@ -230,6 +230,13 @@ static void cb_splunk_flush(const void *data, size_t bytes,
                             ctx->auth_header, flb_sds_len(ctx->auth_header));
     }
 
+    /* Append Channel identifier header */
+    if (ctx->channel) {
+        flb_http_add_header(c, FLB_SPLUNK_CHANNEL_IDENTIFIER_HEADER,
+                            strlen(FLB_SPLUNK_CHANNEL_IDENTIFIER_HEADER),
```
Comment (edsiper) [inline plugins/out_splunk/splunk.c:236]: here you can do ```sizeof(FLB_SPLUNK_CHANNEL_IDENTIFIER_HEADER) - 1``` since is not a variable length

**Review by edsiper — CHANGES_REQUESTED** [review]

**Review by edsiper — APPROVED** [review]

File: plugins/out_splunk/splunk.c (hunk 1 above)
Comment (nokute78) [reply plugins/out_splunk/splunk.c:236]: OK. Thank you.

**Pull request merged**